### PR TITLE
feature(KTP-767): Implemented XGBoost Multioutput Arctan Loss

### PR DESCRIPTION
**Pull request by egordm**
Reference implementation: https://github.com/LaurensSluyterman/XGBoost_quantile_regression/tree/master

Improvements:
* Shorter training time
* Better quantiles
* Similar performance as xgboost_quantile

image

image

### CODE REVIEW

File: openstef/metrics/metrics.py [hunk 1]
```diff
@@ -431,3 +431,54 @@ def xgb_quantile_obj(
...
+    where each row corresponds to the 10 predicted quantile for a single data
+    point. We then use a for-loop (over the 10 columns) to calculate the gradients
+    and second derivatives. Legibility was chosen over efficiency. This part
+    can be made more efficient.
+
+    Parameters:
+        y_true: An array containing the true observations.
+        y_pred: An array containing the predicted quantiles
+        taus: A list containing the true desired coverage of the quantiles
+        s: A smoothing parameter
```
Comment (bartpleiter) [inline openstef/metrics/metrics.py:451]:
```suggestion
        y_pred: An array containing the predicted quantiles.
        taus: A list containing the true desired coverage of the quantiles.
        s: A smoothing parameter.
```

File: openstef/metrics/metrics.py
Comment (bartpleiter) [inline openstef/metrics/metrics.py:1]: I see some inconsistencies with using a full stop after comments/lines in the docstring. I know it is only a minor detail and was already in the example implementation, but I think it is nice if we are consistent with our full stop/period usage :)

File: openstef/model/objective.py [hunk 2]
```diff
@@ -349,6 +349,36 @@ def get_pruning_callback(self, trial: optuna.trial.FrozenTrial):
         )
 
 
+class XGBMultioutputQuantileRegressorObjective(RegressorObjective):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.model_type = MLModelType.XGB_QUANTILE
+
+    def get_params(self, trial: optuna.trial.FrozenTrial) -> dict:
+        """Get parameters for XGBQuantile Regressor Objective with objective specific parameters.
```
Comment (bartpleiter) [inline openstef/model/objective.py:358]:
```suggestion
        """Get parameters for XGB Multioutput Quantile Regressor Objective with objective specific parameters.
```

File: openstef/model/objective.py
Comment (bartpleiter) [inline openstef/model/objective.py:1]: Nice to add the smoothing parameter to the hyperparameter search as Laurens suggested!

File: openstef/model/regressors/xgb_multioutput_quantile.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __init__(
+        self,
+        quantiles: tuple[float, ...] = DEFAULT_QUANTILES,
+        gamma: float = 0.0,
+        colsample_bytree: float = 1.0,
+        subsample: float = 1.0,
+        min_child_weight: int = 0,
+        max_depth: int = 6,
+        learning_rate: float = 0.22,
```
Comment (bartpleiter) [inline openstef/model/regressors/xgb_multioutput_quantile.py:68]: Given the last point in the docstring, would'nt it be better to use 0.05 as learning rate?

File: test/unit/model/regressors/test_xgb_multioutput_quantile.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        with self.assertRaises(ValueError):
+            model = XGBMultiOutputQuantileOpenstfRegressor((0.2, 0.3, 0.5, 0.6, 0.7))
+            model.predict("test_data", quantile=0.8)
+
+    def test_set_params(self):
+        # Check hyperparameters are set correctly and do not cause errors
+
+        model = XGBMultiOutputQuantileOpenstfRegressor((0.2, 0.3, 0.5, 0.6, 0.7))
+
+        hyperparams = {
```
Comment (bartpleiter) [inline test/unit/model/regressors/test_xgb_multioutput_quantile.py:81]: Maybe good idea to add a check for the smoothing parameter as well so we know for sure it is being used

**Review by bartpleiter — APPROVED** [review]
Very  nice! Made some small suggestions

File: openstef/model/objective.py (hunk 2 above)
Comment (egordm) [reply openstef/model/objective.py:358]: Good catch

File: openstef/metrics/metrics.py
Comment (egordm) [reply openstef/metrics/metrics.py:1]: It does look nicer indeed, but in google docstring standard they also mix dot endings and not so it's kinda hard to know or check which is right.

I have changed the args with two more words to use .

File: test/unit/model/regressors/test_xgb_multioutput_quantile.py (hunk 4 above)
Comment (egordm) [reply test/unit/model/regressors/test_xgb_multioutput_quantile.py:81]: Nice suggestion.
